### PR TITLE
[QnAMaker]Making Max and Min ScoreForLowScoreVariation Configurable

### PR DESCRIPTION
**Pull request by vipeketi**
ActiveLearningUtils.MinimumScoreForLowScoreVariation
ActiveLearningUtils.MaximumScoreForLowScoreVariation
QnAMakerDialog.maximumScoreForLowScoreVariation
Made configurable
Issue #3706

### CODE REVIEW

File: libraries/Microsoft.Bot.Builder.AI.QnA/Dialogs/QnAMakerDialog.cs [hunk 1]
```diff
@@ -289,6 +289,8 @@ public QnAMakerDialog([CallerFilePath] string sourceFilePath = "", [CallerLineNu
         /// <seealso cref="RankerTypes"/>
         [JsonProperty("rankerType")]
         public StringExpression RankerType { get; set; } = new StringExpression(RankerTypes.DefaultRankerType);
+      
+        public float MaximumScoreForLowScoreVariation { get => maximumScoreForLowScoreVariation; set => maximumScoreForLowScoreVariation = value; }
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.AI.QnA/Dialogs/QnAMakerDialog.cs:293]: >MaximumScoreForLowScoreVariation [](start = 21, length = 32)

missing comments
missing JsonProperty attribute
should probably be a ExpressionProperty<float>

File: libraries/Microsoft.Bot.Builder.AI.QnA/Utils/ActiveLearningUtils.cs [hunk 2]
```diff
@@ -27,9 +22,18 @@ public static class ActiveLearningUtils
...
+        private static double maximumScoreForLowScoreVariation = 95.0;
+
+        /// <summary>
+        /// Minimum Score For Low Variation.
+        /// </summary>
+        private static double minimumScoreForLowScoreVariation = 20.0;
+
+        public static double MaximumScoreForLowScoreVariation { get => maximumScoreForLowScoreVariation; set => maximumScoreForLowScoreVariation = value; }
+        
+        public static double MinimumScoreForLowScoreVariation { get => minimumScoreForLowScoreVariation; set => minimumScoreForLowScoreVariation = value; }
```
[the remark this answers is not in the document]
Comment (vipeketi) [reply libraries/Microsoft.Bot.Builder.AI.QnA/Utils/ActiveLearningUtils.cs:36]: The field was a constant.

File: libraries/Microsoft.Bot.Builder.AI.QnA/Dialogs/QnAMakerDialog.cs (hunk 1 above)
Comment (vipeketi) [reply libraries/Microsoft.Bot.Builder.AI.QnA/Dialogs/QnAMakerDialog.cs:301]: Implemented.

File: libraries/Microsoft.Bot.Builder.AI.QnA/Dialogs/QnAMakerDialog.cs [hunk 3]
```diff
@@ -290,6 +290,16 @@ public QnAMakerDialog([CallerFilePath] string sourceFilePath = "", [CallerLineNu
...
 
+        /// <summary>
+        /// Gets or sets the Maximum Score For LowScoreVariation to use.
+        /// </summary>
+        /// <value>
+        /// Max score for LowScoreVariation.
+        /// </value>
+        /// <seealso cref="RankerTypes"/>
+        [JsonProperty("maximumScoreForLowScoreVariation")]
+        public float MaximumScoreForLowScoreVariation { get => maximumScoreForLowScoreVariation; set => maximumScoreForLowScoreVariation = value; }
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.AI.QnA/Dialogs/QnAMakerDialog.cs:301]: >MaximumScoreForLowScoreVariation [](start = 21, length = 32)

You need to update MicrosoftQnAMAkerDialog.schema to add maximumScoreForLowScoreVariation

File: libraries/Microsoft.Bot.Builder.AI.QnA/Schemas/Microsoft.QnAMakerDialog.schema [hunk 4]
```diff
@@ -110,6 +110,12 @@
             ]
         }
     },
+        "maximumScoreForLowScoreVariation": {
+           "type": "float",
+            "title": "maximumScoreForLowVariation",
+            "description": "Maximum score For Low Variation",
+            "default": 0.95F
+        },
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.AI.QnA/Schemas/Microsoft.QnAMakerDialog.schema:118]: This needs to be in the properties section.  You are adding a property.

File: libraries/Microsoft.Bot.Builder.AI.QnA/Utils/ActiveLearningUtils.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (tomlm) [reply libraries/Microsoft.Bot.Builder.AI.QnA/Utils/ActiveLearningUtils.cs:36]: My point is you are making these settings GLOBAL for ALL QNA MODELS.  I doubt that's the right thing to do.  Get rid of all "static" in this file and THINK about what you are doing.

---
In reply to: [418081432](https://github.com/microsoft/botbuilder-dotnet/pull/3841#discussion_r418081432) [](ancestors = 418081432)

File: libraries/Microsoft.Bot.Builder.AI.QnA/Schemas/Microsoft.QnAMakerDialog.schema (hunk 4 above)
Comment (vipeketi) [reply libraries/Microsoft.Bot.Builder.AI.QnA/Schemas/Microsoft.QnAMakerDialog.schema]: Added as property

**Comment by vipeketi** [comment]
@tomlm please approve.

File: libraries/Microsoft.Bot.Builder.AI.QnA/Utils/ActiveLearningUtils.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (tomlm) [reply libraries/Microsoft.Bot.Builder.AI.QnA/Utils/ActiveLearningUtils.cs:36]: WAS a private constant and is now a PUBLIC STATIC. That means anyone who changes it changes for all instances of someone calling this function.  WHY is this static? Is it truly global, there is no chance that someone would want to call the api with different settings.  This is bad design.

---
In reply to: [418160756](https://github.com/microsoft/botbuilder-dotnet/pull/3841#discussion_r418160756) [](ancestors = 418160756)

[the remark this answers is not in the document]
Comment (tomlm) [reply libraries/Microsoft.Bot.Builder.AI.QnA/Utils/ActiveLearningUtils.cs:36]: public static = > global varibles.  Global variables are a bad idea unless there is a really good reason to do that.

---
In reply to: [429447761](https://github.com/microsoft/botbuilder-dotnet/pull/3841#discussion_r429447761) [](ancestors = 429447761,418160756)

Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.AI.QnA/Utils/ActiveLearningUtils.cs:36]: dear lord no.

File: libraries/Microsoft.Bot.Builder.AI.QnA/Dialogs/QnAMakerDialog.cs (hunk 3 above)
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.AI.QnA/Dialogs/QnAMakerDialog.cs:301]: >float [](start = 15, length = 5)

Make this
 public NumberExpression MaximumScoreForLowScoreVariation {get; set;}

**Comment by tomlm** [comment]
>         private float maximumScoreForLowScoreVariation = 0.95F;

get rid of this.

---
Refers to: libraries/Microsoft.Bot.Builder.AI.QnA/Dialogs/QnAMakerDialog.cs:77 in f5635ed. [](commit_id = f5635ed2fe7d48ecae13e073b48137fb31da2854, deletion_comment = False)

**Comment by tomlm** [comment]
>             if (isActiveLearningEnabled && response.Answers.Any() && response.Answers.First().Score <= maximumScoreForLowScoreVariation)

var maximumScoreForLowScoreVariation = MaximumScoreForLowScoreVariation.GetValue(dc.State);

---
Refers to: libraries/Microsoft.Bot.Builder.AI.QnA/Dialogs/QnAMakerDialog.cs:452 in f5635ed. [](commit_id = f5635ed2fe7d48ecae13e073b48137fb31da2854, deletion_comment = False)

File: libraries/Microsoft.Bot.Builder.AI.QnA/Schemas/Microsoft.QnAMakerDialog.schema [hunk 5]
```diff
@@ -77,6 +77,12 @@
                 }
             }
         },
+        "maximumScoreForLowScoreVariation": {
+           "type": "float",
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.AI.QnA/Schemas/Microsoft.QnAMakerDialog.schema:81]: >  "type": "float", [](start = 9, length = 18)

     "$ref": "[https://raw.githubusercontent.com/microsoft/botframework-sdk/master/schemas/component/definitions.schema#/definitions/numberExpression](https://raw.githubusercontent.com/microsoft/botframework-sdk/master/schemas/component/definitions.schema#/definitions/numberExpression)",

File: libraries/Microsoft.Bot.Builder.AI.QnA/Utils/ActiveLearningUtils.cs [hunk 6]
```diff
@@ -38,8 +38,31 @@ public static class ActiveLearningUtils
         /// <returns>List of filtered qnaSearch results.</returns>
         public static List<QueryResult> GetLowScoreVariation(List<QueryResult> qnaSearchResults)
         {
-            var filteredQnaSearchResult = new List<QueryResult>();
+            return GetLowScoreVariation(qnaSearchResults, 95);
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.AI.QnA/Utils/ActiveLearningUtils.cs:41]: >95 [](start = 58, length = 2)

return GetLowScoreVariation(MaximumScoreForLowScoreVariation, MinimumScoreForLowScoreVariation,PreviousLowScoreVariationMultiplier,MaxLowScoreVariationMultiplier)

**Comment by vipeketi** [comment]
Abandoning for design change.

**Pull request closed without merging**